lib/handsoap/service.rb

```diff
@@ -96,9 +96,10 @@ class Service
     @@create_document_callback = nil
     @@log_header_callback = nil
     @@log_body_callback = nil
-    attr_reader :http_client
+    attr_reader :http_client, :cookie
     def initialize(ep)
       @http_client = nil
+      @cookie = nil
```

Comment:
Returning a cookie through an instance variable is not thread safe - given that multiple `dispatch` calls can be active at once.

For your specific use case, I think it's OK because the code should be single threaded until the connection is established (however we should still verify that that is indeed the case). The danger here is that if the `cookie` accessor were to be used for any other purpose, it's value may be inconsistent.

I would comment this code to make sure that this is clear, and it's only intended to be used by `login`. An alternative would be to refactor `dispatch` into `raw_dispatch` and `dispatch`. The `raw_dispatch` method would return the whole `response`, including the headers. The `login` method would call `raw_dispatch` and extract the cookie from the headers. The `dispatch` method would call `raw_dispatch` but return `Response.new(response.content, envelope_namespace)` as it currently does.